### ISSUE TITLE
AVMetadataItemValueRequest & SDWebImageManager.load hang 

### ISSUE THREAD

**Issue by iDevPro** [SDWebImage/SDWebImage#3457]
### New Issue Checklist

* [ X] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [ X] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [ X] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | ios
 Platform Version        | 14.0 - 16.2
 SDWebImage Version      | 5.12.1 
 Integration Method      |  cocoapods
 Xcode Version           | 14.2
 Repro rate              | all the time (100%)
 Repro with our demo prj | don't know yet
 Demo project link       | none

### Issue Description and Steps
- iOS application with AVKit, SDWebImage (cocoapods)
- Lazy load artwork like that:
```
extension AVMetadataItem {
    static func artwork(url: URL, placeholder: UIImage?) -> AVMetadataItem {
        let art = AVMutableMetadataItem()
        art.identifier = .commonIdentifierArtwork
        art.dataType = kCMMetadataBaseDataType_PNG as String
        art.extendedLanguageTag = "und"

        return AVMetadataItem(propertiesOf: art) { request in 
              SDWebImageManager.shared
                .loadImage(with: url, context: [:], progress: nil) { image, _, error, _, _, _ in
                     // no call here at all, hang application before this callback
                    let cover: UIImage?

                    if let image = image {
                        cover = image

                    } else {
                        cover = placeholder
                    }

                    guard let data = cover?.pngData() else {
                        request.respond(error: CustomError.loadImageFailed)
                        return
                    }

                    request.respond(value: data as NSData)
                }
        }
}
```

I found some place in code there is hang started in file `SDImageCache.m`:

```
if (doneBlock) {
                if (shouldQueryDiskSync) {
                    doneBlock(diskImage, diskData, SDImageCacheTypeDisk);
                } else {
                    dispatch_async(dispatch_get_main_queue(), ^{ // <<< HERE
                        doneBlock(diskImage, diskData, SDImageCacheTypeDisk);
                    });
                }
            }
```

This hang happen only when call load image in AVMetadataItem function for lazy load, nothing hangs when load by URLSession.shared
Can't yet understand any reason for that :(

**Comment by dreampiggy** [comment]
> This hang happen only when call load image in AVMetadataItem function for lazy load, nothing hangs when load by URLSession.shared

Does your AVKit X SDWebImage use URLSession ? I think it will use AVKit's own networking stack.

Any demo or did you find the reason why this cause the queue lock (or what's exactly that `hang` means 😂)

**Comment by dreampiggy** [comment]
I think you can try to use the `Custom Loader` feature of SDWebImage for your use case, not just use a help method.

See: https://github.com/SDWebImage/SDWebImage/wiki/Advanced-Usage#custom-loader-50

And see how I get PhotosKit X SDWebImage using the custom loader: https://github.com/SDWebImage/SDWebImagePhotosPlugin

**Comment by dreampiggy** [comment]
Or, seems you use 5.12.1. Does the latest version 5.13.3 solve the problem ?

**Comment by iDevPro** [comment]
> > This hang happen only when call load image in AVMetadataItem function for lazy load, nothing hangs when load by URLSession.shared
> 
> Does your AVKit X SDWebImage use URLSession ? I think it will use AVKit's own networking stack.
> 
> Any demo or did you find the reason why this cause the queue lock (or what's exactly that `hang` means 😂)

Demo is this code ;)

**Comment by iDevPro** [comment]
> Or, seems you use 5.12.1. Does the latest version 5.14.2 solve the problem ?

I didn't try this version yet.

**Comment by dreampiggy** [comment]
I guess this is because that `return AVMetadataItem(propertiesOf: art) { request in }`

This completion block, is designed to be called in **async**. However, SDWebImageManager, when image memory cache hit (use using `QueryDiskSync` and hit disk cache), will callback the completion block in **sync**.

This queue issue may cause problem. You'd better provide a full function demo (can run on iPhone Simulator or Real iPhone). And let me investigate. I think for this issue the debugger can pause on the correct timing and see there are two dispatch queue waiting each other.

**Comment by iDevPro** [comment]
[SDWebAVKitMetadataHangs.zip](https://github.com/SDWebImage/SDWebImage/files/10324258/SDWebAVKitMetadataHangs.zip)

**Comment by dreampiggy** [comment]
I think this is because that the SDImageCache's callback queue is `main queue`.

However, the main queue is blocked by `AVFCore`-[AVLazyValueLoadingMetadataItem _waitForLoadingOfValueDependentKey:]:`

The caller queue is`__84-[AVLazyValueLoadingMetadataItem loadValuesAsynchronouslyForKeys:completionHandler:]_block_invoke_3 ()`

Did you get it ?

The problem occurs because:

1. AVKit `main queue` wait for `queue 8`
2. `queue 8` call `SDWebImageManager` to query cache
3. query cache finished, callback to `main queue`
4. dead here, becaue `main queue` wait `queue 8` wait `main queue`

<img width="1110" alt="截屏2022-12-30 21 21 10" src="https://user-images.githubusercontent.com/6919743/210074563-4d181508-d0c2-4265-9306-38f6c5799335.png">
<img width="450" alt="截屏2022-12-30 21 21 17" src="https://user-images.githubusercontent.com/6919743/210074567-7aabe94a-5cd0-4f13-8c40-dcdf40ea46f1.png">

**Comment by dreampiggy** [comment]
The only possible solution is to avoid SDImageCache which always callback to main queue, instead, which should callback to `caller queue`.

Which means, if you call `SDImageCache` in `queue 8`, we callback in `queue 8`, so that it does not blocked by main queue waiting.


But actually, I think this is AVKit design bug or you usage bug. Why main queue wait for a background queue return ?

**Comment by iDevPro** [comment]
"I think you should read AVKit API again ? Is not usage wrong ?)"

**Comment by dreampiggy** [comment]
I found Kingfisher has something feature like this :

https://github.com/onevcat/Kingfisher/blob/70ef0ba088a26561bfee1a2a15efc157696ce33e/Sources/Utility/CallbackQueue.swift

```swift
    /// Sets the callback queue which is used as the target queue of dispatch callbacks when retrieving images from
    ///  cache. If not set, Kingfisher will use main queue for callbacks.
    /// - Parameter queue: The target queue which the cache retrieving callback will be invoked on.
    /// - Returns: A `Self` value with changes applied.
    ///
    /// - Note:
    /// This option does not affect the callbacks for UI related extension methods or `KFImage` result handlers.
    /// You will always get the callbacks called from main queue.
    public func callbackQueue(_ queue: CallbackQueue) -> Self {
        options.callbackQueue = queue
        return self
    }
```

**Comment by dreampiggy** [comment]
> AVMetadataItem

I means, the behavior or API design is that, suitable ?

As AVKit, you design a API, which perform on main queue wait for the async result.

So, the user, who do the async fetch, **should never submit block to main queue**, right ? Because main queue is already been block by AVKit.

**Comment by iDevPro** [comment]
By the way I suppose it is only one place (for me) where I can't use SD :) At now I go to image and get it by URLSession, without SD. Without problems also. And I assume there's not that place for some caching image :)

**Comment by dreampiggy** [comment]
Yes. SDWebImage currently hardcode to callback to main queue, this cause issue.

URLSession's callback queue is control by you. By defaults is just callback in a random global queue, if you don't dispatch it into main queue (and block by AVKit), nothinig worng happended

**Comment by dreampiggy** [comment]
> I found Kingfisher has something feature like this :
> 
> https://github.com/onevcat/Kingfisher/blob/70ef0ba088a26561bfee1a2a15efc157696ce33e/Sources/Utility/CallbackQueue.swift
> 
> ```swift
>     /// Sets the callback queue which is used as the target queue of dispatch callbacks when retrieving images from
>     ///  cache. If not set, Kingfisher will use main queue for callbacks.
>     /// - Parameter queue: The target queue which the cache retrieving callback will be invoked on.
>     /// - Returns: A `Self` value with changes applied.
>     ///
>     /// - Note:
>     /// This option does not affect the callbacks for UI related extension methods or `KFImage` result handlers.
>     /// You will always get the callbacks called from main queue.
>     public func callbackQueue(_ queue: CallbackQueue) -> Self {
>         options.callbackQueue = queue
>         return self
>     }
> ```

SDWebImage introduce the similiar API can resolve this issue. A quick solution is like here:

```objective-c
    /// Sets the callback queue which is used as the target queue of dispatch callbacks when retrieving images from
    ///  cache. If not set, SDWebImage will use main queue for callbacks.
    /// This option does not affect the callbacks for UI related extension methods or `UIView.sd_setImage` result handlers.
    /// You will always get the callbacks called from main queue.
SDWebImageContextOption _Nonnull SDWebImageContextCallbackQueue;
```

**Comment by dreampiggy** [comment]
Use the new API's either `globalQueue` or `callerQueue` can all solve your problem. Actually any queue is not main will rescue the status.

**Comment by dreampiggy** [comment]
Please have a check at #3465.

I've rename some API to make it clear. For example, use `currentQueue` instead of `callerQueue` to make it more clear (By the time you create SDCallbackQueue)

**Comment by iDevPro** [comment]
> Use the new API's either `globalQueue` or `callerQueue` can solve your problem. Actually any queue is not main will rescue the status.
> 
> API usage in Swift
> 
> ```swift
> SDWebImageManager.shared.loadImage(with: url, options: [], context: [.callbackQueue: SDCallbackQueue.global]) in { ... }
> ```

Which version has this API ? I did install 5.14.3 and could not found this key: `.callbackQueue` in ContextOptions..
Is it right that API in `master` or not published yet ?

**Comment by iDevPro** [comment]
<img width="185"  src="https://user-images.githubusercontent.com/1888230/212838241-8449e9a4-2abd-4ccb-80a1-bf3cf6172b3c.png">
<img width="832" src="https://user-images.githubusercontent.com/1888230/212838324-bf60c055-b710-47b1-ba94-6945aca16073.png">

**Comment by dreampiggy** [comment]
[5.15.0](https://github.com/SDWebImage/SDWebImage/releases/tag/5.15.0) released. See changelog for usage.

**Comment by iDevPro** [comment]
Hello @dreampiggy I have checked this code with `SPM + master branch`. Nice fix :)
